daemon.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH daemon 1
.SH NAME
daemon \- run a process as a daemon
.SH SYNOPSIS
.B daemon
.RI [ options ]
.RB [ -- ]
.IR command ...
.SH DESCRIPTION
\fBdaemon\fR runs \fIcommand\fR as a daemon.  It performs the
following steps:
.TP
.B .
All file descriptors above 2 are closed (unless \fB--no-close\fR was
specified)
.TP
.B .
The process switches to the root directory (unless \fB--no-chdir\fR
was specified).
.TP
.B .
If the \fB-l\fR or \fB-L\fR options were used, then a daemon is
created to handle the logging.  If \fB-l\fR was used then any further
errors will go to the logfile.
.TP
.B .
The process forks.  The parent waits for the child to terminate and
then exits (normally this happens almost immediately).
.TP
.B .
In the child, the remaining file descriptors (0, 1 and 2) are
redirected to \fB/dev/null\fR unless the appropriate \fB-l\fR/\fB-L\fR
option or \fB--no-close\fR was used.
.IP
If you didn't specify \fB-l\fR or \fB--no-close\fR then any errors
from here on will not be reported anywhere.
.TP
.B .
The process calls \fBsetsid\fR(2) to create a new session.
.TP
.B .
The process forks again, so that it isn't the session leader.  The
parent exits immediately.
.TP
.B .
In the child, the command is executed.
.SH OPTIONS
.TP
\fB-n\fR, \fB--no-close\fR
Don't close file descriptors.
.TP
\fB-l\fR \fIpath\fR, \fB--log-stderr\fR \fIpattern\fR
Log stderr to a file.  \fIpatter\fR has the same syntax as used by
\fBstrftime\fR(3).
.TP
\fB-L\fR \fIpath\fR, \fB--log-stdout\fR \fIpattern\fR
Log stdout to a file.
.TP
\fB-C\fR, \fB--no-chdir\fR
Don't change directory.
.TP
\fB-m\fR \fIN\fR, \fB--max-log-age\fR \fIN\fR
Delete logfiles that are older than \fIN\fR days old.  By default,
logfiles are kept forever.
.TP
\fB-c\fR, \fB--compress\fR
Compress saved logfiles.
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>.


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



daemon.c
/*
   daemon - run program as a daemon

   Copyright (C) 2001, 2002 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "utils.h"
#include "logdaemon.h"

static int closefds = 1;
static int changedir = 1;

/* Option flags and variables */

static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { "no-close", no_argument, 0, 'n' },
  { "no-chdir", no_argument, 0, 'C' },
  { "log-stderr", required_argument, 0, 'l' },
  { "log-stdout", required_argument, 0, 'L' },
  { "compress", no_argument, 0, 'c' },
  { "max-log-age", required_argument, 0, 'm' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fputs(
"Usage:\n"
"  daemon [options] [--] command ...\n"
"\n"
"Options:\n"
"  -C, --no-chdir                        Don't \"cd /\"\n"
"  -n, --no-close                        Don't close file descriptors\n"
"  -h, --help                            Usage message\n"
"  -l PATTERN, --log-stderr PATTERN      Log standard error to a file\n"
"  -L PATTERN, --log-stdout PATTERN      Log standard output to a file\n"
"  -c, --compress                        Compress logs\n"
"  -m DAYS, --max-log-age DAYS           Delete old logs\n"
"  -V, --version                         Version number\n"
, fp) < 0)
    fatale("output error");
  exit(exit_status);
}

int main(int argc, char **argv) {
  int n;
  int nullfd;
  int pipefds[2];
  pid_t pid;
  int max = 0;
  int compress = 0;
  struct sigaction sa, osa;
  const char *logs[3] = { 0, 0, 0 };

  setprogname(argv[0]);

  while((n = getopt_long(argc, argv,
			 "hVncl:m:CL:",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("daemon %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);
      return 0;

    case 'n':
      closefds = 0;
      break;

    case 'C':
      changedir = 0;
      break;

    case 'm':
      max = atoi(optarg);
      break;

    case 'c':
      compress = 1;
      break;

    case 'l':
      logs[2] = optarg;
      break;

    case 'L':
      logs[1] = optarg;
      break;

    default:
      usage(stderr, 1);
    }
  }

  if(optind >= argc)
    fatal("no command specified");

  /* zap unwanted file descriptors (before we open log pipes!) */
  if(closefds) {
    for(n = maxfd(); n > 2; --n)
      close(n);
  }

  if(changedir) {
    /* change directory */
    if(chdir("/") < 0)
      fatale("error calling chdir");
  }

  if(logs[1] || logs[2]) {
    /* set up a logging daemon straight away */
    int p[3][2];
    struct logfile *l[3];
    int fd;

    /* create the pipes */
    for(n = 1; n <= 2; ++n)
      if(logs[n]) {
	pipe_e(p[n]);
	l[n] = ld_new_logfile(logs[n]);
	l[n]->rotate = max;
	l[n]->compress = compress;
	ld_new_input(p[n][0], l[n]);
      }
    /* create the daemon */
    if(!(pid = fork_e())) {
      exiter = _exit;
      setsid_e();
      for(n = 1; n <= 2; ++n)
	if(logs[n])
	  close_e(p[n][1]);
      fd = open_e("/dev/null", O_RDWR, 0);
      /* zap the real stdin/stdout/stderr */
      for(n = 0; n < 3; ++n)
	dup2_e(fd, n);
      if(fd >= 3)
	close_e(fd);
      if(!fork_e())
	ld_loop();
      _exit(0);
    }
    /* close the input ends now */
    for(n = 1; n <= 2; ++n)
      if(logs[n])
	close_e(p[n][0]);
    /* wait for the intermediate subprocess to terminate before
     * slamming stderr */
    waitpid_e(pid, &n, 0);
    if(n)
      fatal("logger subprocess %s", wstat(n));
    /* redirect stdout and/or stderr to the pipes */
    for(n = 1; n <= 2; ++n)
      if(logs[n] && p[n][1] != n) {
	dup2_e(p[n][1], n);
	close_e(p[n][1]);
      }
  }

  /* first fork... */
  if((pid = fork_e())) {
    /* wait for the intermediate subprocess (that will become the new
     * session leader) to terminate and notify the caller of any error
     * via our exit status */
    exiter = _exit;
    waitpid_e(pid, &n, 0);
    if(n)
      fatal("subprocess %s", wstat(n));
    _exit(0);
  }

  /* Stevens says the child will get a SIGHUP when the session leader
   * terminates, so we pre-emptively ignore it.  We'll restore it when
   * we know that the session leader has gone. */
  sa.sa_handler = SIG_IGN;
  sigemptyset(&sa.sa_mask);
  sa.sa_flags = 0;
  sigaction_e(SIGHUP, &sa, &osa);

  /* in the first child */

  /* create a pipe so that the ultimate child knows when we
   * terminate */
  pipe_e(pipefds);

  /* zap stdin/stdout/stderr */
  if(closefds) {
    nullfd = open_e("/dev/null", O_RDWR, 0);
    for(n = 0; n <= 2; ++n)
      if(!logs[n] && nullfd != n)
	dup2_e(nullfd, n);
    if(nullfd > 2)
      close_e(nullfd);
  }

  /* create a new session */
  setsid_e();

  /* now we're a session leader.  But this is bad, as it means if we
   * ever open a tty device, it'll become our controlling terminal,
   * and we'll be liable to SIGHUPs from it. */

  if(fork_e())
    /* parent - terminate immediately */
    _exit(0);

  /* child - wait for the pipe to close so that we know when we can
   * restore SIGHUP */
  {
    char buffer[1];
    close_e(pipefds[1]);
    do {
      n = read(pipefds[0], buffer, 1);
    } while(n < 0 && errno == EINTR);
    if(n < 0)
      fatale("error calling read");
    close_e(pipefds[0]);
    sigaction_e(SIGHUP, &osa, 0);
  }

  /* now we are detached */

  /* now all that's left to do is execute the daemon program itself */
  execvp(argv[optind], argv + optind);

  fatale("error calling execvp %s", argv[optind]);
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
